Proposed patches to add IA-16 far pointer support

tkchia opened the pull request:
My proposed patches for `gcc-ia16` to add some (very preliminary) support for segment:offset far pointers.  With the patches, I can now write something like this:

```c
int main(void)
{
	static const char msg[] = " -*- Hello world! -*- ", *p;
	volatile unsigned __far *q = (unsigned __far *)0xb8000000ul;
	for (p = msg; *p; ++p)
		*q++ = 0x0f00u | (unsigned char)*p;
	return 0;
}
```

As I am still not very familiar with `gcc` internals, I suspect the patches will need a lot more improvement and testing to pass muster, but I am submitting them in case they might be useful.

tkchia committed:
Add very preliminary far pointer support (take 2)
(while still trying to learn about gcc internals...)

tkchia committed:
Far pointer support: dereferencing far pointers now seems to work (-:

tkchia committed:
Now no longer tries to spew bogus addresses like -1(%bx,%bp) --
tweaked gcc/regcprop.c's replace_oldest_value_addr(...) to
signal the presence of an index register down to its recursive
subcalls

tkchia committed:
Slightly streamline ia16_parse_address (...)

tkchia committed:
Stop gcc/regcprop.c from munging (unspec ...) segment terms in addresses

reenigne commented:
Thank you for this - great work! I had made a start on this feature myself (I just pushed it to https://github.com/crtc-demos/gcc-ia16/tree/gcc-6_3_0-ia16-address_spaces so that you can compare), but you got further than I did since yours actually works! As you suspected, your patches do still need some work, but it looks like there's enough of it here to be useful. I have some questions before I merge it, though:
1) Do you have (or can you get) a copyright assignment (either for this change specifically, or a more general one if you'd like to make further contributions in the future)? I may try to get this port upstream, in which case copyright assignments will be needed from all contributors.
2) Why is the the new argument to ia16_parse_address() (and elsewhere) called p_r9? I'd have thought p_sr (for segment register) or p_rs would be a better name (as you've used elsewhere).
3) Why is the hack in explow.c necessary? It's better to avoid making changes to GCC core code (i.e. outside of gcc/config/ia16) if at all possible, but when it is necessary it's important to make it as generic as possible, well-documented and with testcases where appropriate. It's probably not appropriate to put "Very preliminary IA-16 far pointer support by TK Chia" at the top of these files since they don't have such notices for all the other changes people have made to them (this is fine for the backend files though). Likewise with the changes in regcprop.c.
4) In lra-constraints.c, do you think the debugging information could be dumped to the standard debugging output instead of guarded with "#if 0"?
5) At the moment, the only register in the SEGMENT_REGS class is ES. It would be good to add DS as well so that loops which access far pointers in two segments don't necessarily have to load a new value into ES twice per iteration. Of course, this adds complexity for accessing near addresses other than [bp+offset] and [bp+index+offset]. Any thoughts on how to do this?

Thanks again!

tkchia committed:
Removed extraneous author notices from gcc/explow.c,
gcc/lra-constraints.c, gcc/regcprop.c, gcc/rtl.h

tkchia commented:
Thank you very much for your feedback!

> Do you have (or can you get) a copyright assignment (either for this change specifically, or a more general one if you'd like to make further contributions in the future)? I may try to get this port upstream, in which case copyright assignments will be needed from all contributors.

Yes.  The work was done in my own time; and I am willing put these specific changes in the public domain so that the FSF may use them.

[_update 17 Jul 2017:_ I am also willing to provide a copyright assignment to the FSF, if this is possible.]

> Why is the the new argument to ia16_parse_address() (and elsewhere) called p_r9? I'd have thought p_sr (for segment register) or p_rs would be a better name (as you've used elsewhere).

That was merely for symmetry with `p_r1` and `p_r2` (in contrast to p_ri and p_rb).  But `p_rs` works for me too.

> Why is the hack in explow.c necessary? It's better to avoid making changes to GCC core code (i.e. outside of gcc/config/ia16) if at all possible, but when it is necessary it's important to make it as generic as possible, well-documented and with testcases where appropriate.

I did the `explow.c` hack because I found that the IA-16 backend code was not given a chance to "legitimize" certain far pointer expressions before `explow.c` decided that they were invalid.

Specifically, when `memory_address_addr_space(`...`)` saw an expression like `(void __far *)0x0040006c` -- which would become RTX `(reg:SI` ...`)` -- it would not try to invoke `TARGET_ADDR_SPACE_LEGITIMIZE_ADDRESS` first.  Instead it would go straight to the `gcc_assert (memory_address_addr_space_p (`...`));` which would fail.

I have not quite found another way to properly munge the far address without this hack.

I will have to explain my changes to `regcprop.c` and `lra-constraints.c` later.

> It's probably not appropriate to put "Very preliminary IA-16 far pointer support by TK Chia" at the top of these files since they don't have such notices for all the other changes people have made to them (this is fine for the backend files though). Likewise with the changes in regcprop.c.

_Mea culpa_.  I will take out the notices.

> In lra-constraints.c, do you think the debugging information could be dumped to the standard debugging output instead of guarded with "#if 0"?

I believe so.

> Of course, this adds complexity for accessing near addresses other than [bp+offset] and [bp+index+offset]. Any thoughts on how to do this?

I will try to think about this. :)  One problem I see is that the current ABI assumes `%ds` points to the generic data space -- if a function can change `%ds` _and_ it also calls other functions, then we need to either break the ABI somehow, or arrange for `%ds` to be restored before each lower-level call.

Thank you again!

tkchia commented:
The changes to `gcc/regcprop.c` -- which I realize might intrude on the workings of other `gcc` ports -- were because I found that, when confronted with an address RTX along the lines of
```
                (plus:HI)
                  /  \
          (plus:HI)  (reg:HI 10 bp)
            /  \
(reg:HI 9 di)  (const_int -1)
```
the copy propagator would see the `%di + (-1)` expression and decide it was OK to explace with `%di` with `%bx`, and then it decided not to replace `%bp`, leading to a bogus address `-1(%bx,%bp)`.  This problem _did_ affect code that were not using far pointers -- including `gcc`'s `sound.c` test case.

I think this problem arose, partly because I rewrote `ia16_parse_address_internal(`...`)` to be more permissive and accept any permutation of { base, index, displacement, segment }.  I see that Lambertsen's original code insisted that the base and index must come immediately under the same `(plus` ...`)`.

So perhaps one way to avoid hacking `regcprop.c` would be for me to make `ia16_parse_address_internal(`...`)` a bit less permissive again -- please let me know what you think.

tkchia commented:
The changes to `gcc/lra-constraints.c` just serve to make sure that the inner expression inside each `(unspec` ...`)` segment expression is reloaded into an appropriate register.  It only comes into play if the back-end code defines the macro `MODE_SEGMENT_REG_CLASS (`...`)` (a name I made up), so it should not intrude on the workings of the other compiler ports.

tkchia committed:
Make ia16_parse_address_internal (...) less permissive --
now requires base and index registers, if present, to come
immediately under the same (plus ...) RTX; this is so that I
can remove most of the intrusive hacks to gcc/regcprop.c

reenigne commented:
> Yes. The work was done in my own time; and I am willing put these specific changes in the public domain so that the FSF may use them.

Please follow the instructions at https://gcc.gnu.org/contribute.html to get your copyright assignment documentation on file.

> I will try to think about this. :) One problem I see is that the current ABI assumes %ds points to the generic data space -- if a function can change %ds and it also calls other functions, then we need to either break the ABI somehow, or arrange for %ds to be restored before each lower-level call.

I think the ABI should remain the same and %ds should be set to %ss before calling another function if it was changed (unless we're using the huge memory model, in which case the child function will set up its own value for %ds).

> I think this problem arose, partly because I rewrote ia16_parse_address_internal(...) to be more permissive and accept any permutation of { base, index, displacement, segment }. I see that Lambertsen's original code insisted that the base and index must come immediately under the same (plus ...).

Ah, that makes sense. I tried to make a similar change myself at one point and had similar problems. I think the right answer is for ia16_parse_address_internal() to not be permissive. I seem to recall reading somewhere that there is a canonical way to arrange the elements of an address RTX, but I can't seem to find where I read that now.

tkchia commented:
> Please follow the instructions at https://gcc.gnu.org/contribute.html to get your copyright assignment documentation on file.

I will work on this.

> I think the right answer is for ia16_parse_address_internal() to not be permissive.

Thank you.  I have pushed another patch which gets `ia16_parse_address_internal(`...`)` to forbid the base-separate-from-index case, and undoes the more intrusive `regcprop.c` hacks.

I will see if I can document and tidy up the rest of the code over the next few days/weeks.

> I think the ABI should remain the same and %ds should be set to %ss before calling another function if it was changed

Do you think it will be worthwhile to define a new kind of function attribute -- similar to x86-32 `__attribute__((regparm))` -- which will control whether a function is compiled with `%ds` made available for use?

>  (unless we're using the huge memory model, in which case the child function will set up its own value for %ds).

I have not yet thought about getting `gcc` to implement a "large" or "huge" memory model. :-)  I am still figuring out how (or if) I can get a `__far` address like `0x1234:0xffff` to wrap around to either `0x1234:0x0000` or `0x2234:0x0000` (not `0x1235:0x0000`), which apparently is non-trivial in itself.  It seems to me that large chunks of `gcc` still assume flat, non-segmented address spaces, with pointers being just plain old integers (of various modes).

reenigne commented:
> I will work on this.
> I think the right answer is for ia16_parse_address_internal() to not be permissive.
Thank you. I have pushed another patch which gets ia16_parse_address_internal(...) to forbid the base-separate-from-index case, and undoes the more intrusive regcprop.c hacks.
> I will see if I can document and tidy up the rest of the code over the next few days/weeks.

Thanks for all that!

> Do you think it will be worthwhile to define a new kind of function attribute -- similar to x86-32 __attribute__((regparm)) -- which will control whether a function is compiled with %ds made available for use?

In terms of having code compiled in one memory model calling code compiled in another, such an attribute could be useful (though users will still have to take care because attributes like these are not recorded in the object file or enforced by the linker). I'm not sure if it'd be worthwhile having a calling convention where %ds is not assumed but the memory model is not huge - calling functions and accessing near memory via %ds is probably going to be much more common than using %ds for far accesses.

> I have not yet thought about getting gcc to implement a "large" or "huge" memory model. :-) I am still figuring out how (or if) I can get a __far address like 0x1234:0xffff to wrap around to either 0x1234:0x0000 or 0x2234:0x0000 (not 0x1235:0x0000), which apparently is non-trivial in itself.

I think 0x1234:0x0000 would be the right thing here. The old DOS compilers treated a far pointer as a near pointer with an extra 16-bit field that doesn't participate in arithmetic. A huge pointer, on the other hand, is renormalized after every arithmetic operation (and you can add a long to a huge pointer and have it do the right thing) so 0x2234:0x0000 would be the correct result for a huge pointer.

> It seems to me that large chunks of gcc still assume flat, non-segmented address spaces, with pointers being just plain old integers (of various modes).

Tricky. It might be necessary to invent new machine modes (with PARTIAL_INT_MODE?) for far and huge pointers instead of using SImode, return these from ia16_addr_space_pointer_mode() and ia16_addr_space_address_mode() and then define how arithmetic works in these modes (not sure how to do that). Actually, it's a weird hack but I wonder if CHImode might work for the far pointer case.

tkchia committed:
Treat __far address 0x0000:0x0000 as valid; and, add
ATTRIBUTE_UNUSED to unused function parameters and variables

tkchia committed:
Add one more argument to MODE_SEGMENT_REG_CLASS -- to make it
less confusing to use

tkchia committed:
Document IA-16 far pointer usage and internals

tkchia committed:
Cost calculations now include segment overrides, at least for i8086/i8088

tkchia committed:
Some test cases for IA-16 far pointers, including a few which
are expected to fail

tkchia committed:
Add another expected-to-fail test case for IA-16 __far

tkchia committed:
IA-16 documentation: note that non-volatile far pointers are not yet working

tkchia commented:
> It might be necessary to invent new machine modes (with PARTIAL_INT_MODE?) for far and huge pointers instead of using SImode, return these from ia16_addr_space_pointer_mode() and ia16_addr_space_address_mode() and then define how arithmetic works in these modes (not sure how to do that). Actually, it's a weird hack but I wonder if CHImode might work for the far pointer case.

Thanks!  I just tried further hacking my far pointer implementation to define and use a new `PSImode`, but found that the semantics quickly got _really_ weird -- for starters,

* both pointers and offsets have to be converted to `PSImode` before they are added;
* there is no really reliable way to tell which of the operands in a `(plus:PSI` ...`)` is a pointer (I did try `MEM_POINTER (.)` and `REG_POINTER (.)`);
* an offset can potentially be negative, e.g. `0xfffe`, which sign-extends to `0xffff'fffe`; ...

My suspicion is that something like `CHImode` will work better in the long run (if it works).  For now, I find that writing e.g.
```
*(volatile unsigned long _far *)(0x046c + 0ui)
```
will already make `gcc` reject the program way before the RTL stages (in `convert_to_pointer_1 (`...`)` in `gcc/convert-c`).  So there is that...

tkchia commented:
> My suspicion is that something like `CHImode` will work better in the long run (if it works).

I might have spoken too soon.  I now think it may be possible to redefine part of the logic for lowering (translating) GIMPLE into RTL --- I am guessing in `gcc/expr.c` (and perhaps elsewhere) --- so that `POINTER_PLUS_EXPR` expressions are not simply mapped to `(plus` ...`)` nodes, and far pointer arithmetic works as "expected".

This should be more reliable than trying to divine the high-level semantics of the low-level `(plus` ...`)` forms.

tkchia committed:
Two more test cases for IA-16 far pointers

tkchia commented:
I have completed the copyright assignment filing.  There should be no legal issues with getting my current patches upstream now.

I am also thinking about how to fix the compiler crashes that happen with non-volatile far pointers.

It seems that part of the problem is dealing some of the conventions and assumptions in the `gcc` source code.  This is my understanding (I hope I am reading the sources right):

  * A "pointer" is the relatively inert thing that a program can stash in memory, load into registers, increment, decrement, etc.
  * An "address" is what happens when the program actually wants to _dereference_ a pointer to the memory it refers to.
  * `Pmode` is actually a mode for an address (!).
  * `gcc` assumes that `sizeof (`_address_`) >= sizeof (`_pointer_`)` within each address space, and that in the worst case one can stash an address in a single register of the correct mode (using `POINTERS_EXTEND_UNSIGNED` and a `ptr_extend` insn).

If I continue with the general approach of my current patches, I will have

  * a "far pointer" being just an `SImode` (or `PSImode`) quantity, except perhaps with weird arithmetic; and
  * a "far address" being a `HImode` which is "coloured" by an `(unspec` ...`)` segment term (so that `gcc` can allot registers independently).

This arrangement is already breaking `gcc`'s assumptions big time -- so it seems I will have to patch up around some of the uses of `POINTERS_EXTEND_UNSIGNED`.

tkchia commented:
As for freeing up `%ds` for use with far pointers, I think one way might be to make near addresses and far addresses have the same "shape" (i.e. both types have segment terms), and then let the compiler optimize how `%ss` is loaded into `%ds` and/or `%es`.

We can perhaps create a special kind of `(unspec` ...`)` segment term which denotes the generic data segment (`%ss`), then add this special term to "addresses" that are derived from near pointers.  But this is still very speculative on my part.

Thanks!

tkchia committed:
IA-16 __far: make test cases more rigorous

tkchia committed:
IA-16 __far: minor cleanup hacks

  * regularized hook routine names in gcc/config/ia16/ia16.c
  * replaced TARGET_VALID_POINTER_MODE hook with
    TARGET_ADDR_SPACE_VALID_POINTER_MODE hook
  * moved address legitimization hack within gcc/explow.c into
    convert_memory_address_addr_space (...), which is used for
    pointer-to-address conversions anyway
  * tweaked addr_for_mem_ref (...) in gcc/tree-ssa-address.c
    to use the address space argument, and to (indirectly)
    legitimize addresses if needed
  * non-volatile __far pointers are still not working  :-(

reenigne commented:
This is all great stuff - thanks again! Sorry I haven't commented on any of it for a while, but I don't really have anything else to add. Now that you've got your copyright assignment filed, I'm happy to merge this pull request whenever you like. Let me know if I should go ahead and do that now or if you have any other changes you want to make to it first. I'm also happy to add you as a collaborator so that you can push directly to it, if you like.

Have you figured out how to build the simulator for running tests and to compare the resulting logs to check for performance regressions? The source for the simulator is at https://github.com/reenigne/reenigne/blob/master/8088/86sim/86sim.cpp (you might need to go back a couple of revisions - I think the current version of the simulator spews a lot of output) and https://github.com/reenigne/reenigne/blob/master/logtools/log_compare/log_compare.cpp is the log compare tool (depends on some of the headers under https://github.com/reenigne/reenigne/tree/master/include). Let me know if you have any problems building/using these. I suspect your changes won't introduce any regressions but it'd be good to check.

tkchia commented:
> Let me know if I should go ahead and do that now or if you have any other changes you want to make to it first.

Thank you!  I now find though that there is a code correctness issue -- the induction variable optimization pass (`.154t.ivopts`) keeps trying to cast 32-bit pointer values to 16-bit `sizetype`.

This affects even the use of far pointers to `volatile` types, so to me the problem seems quite serious.  I think I will try to give a shot at fixing at this issue first before a merge (my apologies).

> Have you figured out how to build the simulator for running tests and to compare the resulting logs to check for performance regressions?

I did try to look for `86sim.cpp` and `log_filter`, but I guess I was looking in the wrong places. :-(  I am currently using `dosemu` to do the tests.

Thank you for the links -- I will try out your tools, and let you know if I have any problems.

reenigne commented:
> This affects even the use of far pointers to volatile types, so to me the problem seems quite serious. I think I will try to give a shot at fixing at this issue first before a merge (my apologies).

Okay, let me know if you get stuck and I'll take a look myself (it'll probably be a few days before I have time though).

> I did try to look for 86sim.cpp and log_filter, but I guess I was looking in the wrong places. 

I should probably move them to the build-ia16 project. The log_filter code is currently at https://github.com/reenigne/reenigne/blob/master/logtools/log_filter/log_filter.cpp - all it does is remove the FAILs caused by test cases that are too big to fit in a .com file (hopefully we'll be able to get rid of it once we've got far pointers, bigger memory models and .exe support working).

tkchia committed:
IA-16 __far: properly distinguish between pointers and addresses,
and allow far addresses to be converted back into far pointers

tkchia committed:
IA-16 __far: suppress iv optimization for far pointers

This fixes a rather bad code correctness issue when a far
pointer is involved in a loop.  The test cases gcc.target/
ia16/far-addr-{7, 8}.c should now pass.

tkchia committed:
IA-16 __far: update write-up -- pointers to non-volatile now work

tkchia committed:
IA-16 __far: fix typo affecting far address -> far pointer conversions

tkchia commented:
I think I have fixed the `ivopts` correctness issue, in a way.  I found a way to suppress the optimization for cases involving far pointers, since it seems hard to get these handled right anyway.

I believe my code should be ready for a merge now.  I tested it on my `gcc.target/ia16/*.c` test cases, and also ran part of the rest of the testsuite.

Meanwhile, I will also see if I can do a more comprehensive test, with the help of your `86sim`. :)

The only really intrusive change I made to the GCC core will be in `addr_for_mem_ref (`...`)` in `gcc/tree-ssa-address.c`.  Perhaps I am missing something, but the original code makes no sense to me (it would sometimes return a pointer, sometimes an address, and when turning pointers into addresses it would discard the address space information).

(The rest of the core code changes are conditioned on `MODE_SEGMENT_REG_CLASS`, which works for the moment, though I suspect that I might be overloading this macro with too many meanings...)

Thank you!

tkchia committed:
IA-16 __far: more rigorous far-addr-7.c test case
(which earlier got lost in a "git rebase")

tkchia committed:
IA-16 __far: beefing up test cases and making them work with 86sim

  * far-addr-2.c now outputs any differing values
  * added dg-excess-errors directives to far-addr-3.c and
    far-addr-6.c
  * plus, far-addr-6.c now also tries reading a near variable
    via a far pointer, after altering it as a near variable
  * far-addr-8.c now only accesses 0xfff0:0x0000--0x00ff
  * this leaves far-addr-5.c, which does not work with 86sim
    as it uses `int $0x1a'

tkchia commented:
> The log_filter code is currently at https://github.com/reenigne/reenigne/blob/master/logtools/log_filter/log_filter.cpp

I have patched ([q.v.](https://github.com/tkchia/build-ia16/blob/c1fed8dc717cdcfc0c4b1f7c5c2b0a290b346134/alfe-ubuntu.diff)) some of the headers used by `log_filter.cpp`, to remove some remaining Visual C++-isms, and to fix some problems warned by Ubuntu `g++`.  Thanks!

tkchia commented:
I did a comparison of the `make check` results over the C test cases (I have not completed the C++ tests) for my patched `gcc`, against the results from `sourceryg++-2016.11-64-ia16-elf.public.tests.tar.bz2` in your April 2017 code release.

To gauge correctness, I did a simple `diff -U2` of `x86_64-linux-gnu-gcc.sum` against my `gcc.sum`.

  * There were no regressions -- cases where a `PASS` became a `FAIL`.
  * However, I see that `gcc.dg/pr52904.c` and `gcc.dg/tree-ssa/forwprop-28.c` were marked `XFAIL` where they were originally a `PASS`.
  * Some cases were originally `FAIL` (apparently due to timing out within `86sim`) but are now `PASS`.
  * My new test cases for far pointers had some problems running under `86sim`.  I have (since) added some patching of `86sim` to initialize part of the interrupt vector table and ROM BIOS area, so most of these tests now work.  I am not sure how to adapt my `gcc.target/ia16/far-addr-5.c` though -- it uses `int $0x1a`.

As for speed and size, I obtained these results from `log_compare`:
```
Speed:
9469 tests got faster by an average of -3316.310804 cycles.
0 tests got slower by an average of -nan cycles.
Cycles: Before 650910.535454, after: 647770.006701 (-0.484822%) in 9999 tests
Size:
1 tests got smaller by an average of -602.000000 bytes.
1773 tests got larger by an average of 79.388607 bytes.
Bytes: Before 2835.780749, after: 2910.729412 (2.574910%) in 1870 tests
```

If you would like to take a look at the original logs, please let me know.  Thank you!

tkchia committed:
IA-16: __builtin_return_address (0) now triggers stack frame setup

gcc/config/ia16/ia16.{c, h} now arrange to set up a stack frame
and frame pointer (%bp) whenever a function invokes
__builtin_return_address (0) --- even if a frame pointer is not
otherwise needed.

Without a frame pointer, __builtin_return_address (0) will not
work properly (the compiled code will try to dereference %bp
without initializing it).

This fixes the failure of gcc/testsuite/gcc.dg/20001117-1.c .

tkchia committed:
Merge commit '39de6b8a' into gcc-6_3_0-ia16-tkchia

tkchia committed:
IA-16: casting between generic and far pointers now works

Also beefed up gcc/testsuite/gcc.target/ia16/far-addr-6.c,
which should now pass.

tkchia committed:
IA-16 __far: now defines a new macro TARGET_ADDR_SPACE_WEIRD_P (as)

tkchia committed:
IA-16: C front-end now defines __FAR macro to indicate __far support

I also beefed up gcc/testsuite/gcc.target/ia16/far-addr-1.c to
test that __FAR is indeed defined.

tkchia committed:
IA-16: separate "any segment register" and "%es" constraints

Before this patch, there was just one assembly operand
constraint "Q" which could mean either "any (available)
segment register" or "the %es register".

This patch adds a new constraint "e".

The intent is that the user should specify "e" if an operand
absolutely must go into %es --- for example because an ABI
mandates it --- and specify "Q" if it is fine for it to go
into (say) %ds, if %ds ever becomes available for allocation.

"e" currently maps internally to the same register class as
"Q", namely SEGMENT_REGS, but this may change.

I have also added two test programs which check that both
"Q" and "e" work as they should.

tkchia commented:
With my latest patches (up to commit `049677d441`), the test result summary for the `gcc` tests is as follows:

```
# of expected passes            76013
# of unexpected failures        50
# of expected failures          112
# of unresolved testcases       22
# of unsupported tests          2665
/home/user/.local/build-ia16/build2/gcc/xgcc  version 6.3.0 (GCC) 
```

The unexpected failures are as follows:

```
FAIL: gcc.c-torture/execute/builtins/memcpy-chk.c compilation,  -O3 -fomit-frame-pointer -funroll-loops -fpeel-loops -ftracer -finline-functions 
FAIL: gcc.c-torture/execute/builtins/memmove-chk.c compilation,  -O3 -fomit-frame-pointer -funroll-loops -fpeel-loops -ftracer -finline-functions 
FAIL: gcc.c-torture/execute/builtins/mempcpy-chk.c compilation,  -O3 -fomit-frame-pointer -funroll-loops -fpeel-loops -ftracer -finline-functions 
FAIL: gcc.c-torture/execute/builtins/memset-chk.c compilation,  -O3 -fomit-frame-pointer -funroll-loops -fpeel-loops -ftracer -finline-functions 
FAIL: gcc.c-torture/execute/builtins/memset-chk.c compilation,  -O3 -g 
FAIL: gcc.dg/compat/scalar-by-value-3 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-10 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-11 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-12 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-13 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-14 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-15 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-16 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-16a c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-17 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-17a c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-18 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-18a c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-22 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-2 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-4 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-5a c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-6a c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-6b c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-7a c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-7b c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-8 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-by-value-9 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/compat/struct-return-2 c_compat_x_tst.o-c_compat_y_tst.o link
FAIL: gcc.dg/long_branch.c (test for excess errors)
FAIL: gcc.dg/loop-9.c scan-rtl-dump loop2_invariant "Decided"
FAIL: gcc.dg/loop-9.c scan-rtl-dump loop2_invariant "without introducing a new temporary register"
FAIL: c-c++-common/torture/builtin-arith-overflow-13.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/builtin-arith-overflow-2.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/builtin-arith-overflow-3.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/builtin-arith-overflow-3.c   -O2  (test for excess errors)
FAIL: c-c++-common/torture/builtin-arith-overflow-3.c   -O2 -flto -fno-use-linker-plugin -flto-partition=none  (test for excess errors)
FAIL: c-c++-common/torture/complex-sign-add.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/complex-sign-mixed-add.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/complex-sign-mixed-mul.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/complex-sign-mixed-sub.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/complex-sign-mul.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/complex-sign-sub.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/vector-compare-1.c   -O0  (test for excess errors)
FAIL: c-c++-common/torture/vector-compare-1.c   -O3 -fomit-frame-pointer -funroll-loops -fpeel-loops -ftracer -finline-functions  (test for excess errors)
FAIL: c-c++-common/torture/vector-compare-1.c   -O3 -g  (test for excess errors)
FAIL: gcc.dg/torture/fp-int-convert-double.c   -O0  (test for excess errors)
FAIL: gcc.dg/torture/fp-int-convert-long-double.c   -O0  (test for excess errors)
FAIL: gcc.dg/tree-ssa/alias-32.c scan-tree-dump-times cddce1 "\\*[^ ]" 1
FAIL: gcc.target/ia16/far-addr-5.c execution test
```

The execution test failure on my `far-addr-5.c` is because (as I previously stated) the test uses `int $0x1a`, which `86sim` does not (yet) understand.

Thank you!

tkchia committed:
IA-16: now with proper far pointer arithmetic ;-)

  * The output code for __far arithmetic is not very
    efficient, for some reason (`movw %es, %dx; movw %dx,
    %es; ...').  My guess is that GCC keeps trying to
    convert from pointers to addresses, then from addresses
    back to pointers, and the conversions are causing all
    the extra code.

    This weirdness probably needs further study.  And, the
    whole feature should be more rigorously tested.

  * In implementing this feature, I also defined (and
    documented) a new TARGET_EXPAND_WEIRD_POINTER_PLUS_EXPR
    internal macro, which is invoked by the machine-
    -independent code if defined.

  * The test case gcc/testsuite/gcc.target/ia16/far-addr-4.c
    now passes.

(cherry picked from commit b898f780151db691275bd22402dec50eed1d7d07)

tkchia committed:
IA-16: peephole-optimize far pointer arithmetic in loops

tkchia committed:
IA-16: beef up far pointer arithmetic tests far-addr-{4, 7}.c

(cherry picked from commit 18ddcb7429b8f19f10ef1b5dfc3d5d31062d7eeb)

tkchia committed:
IA-16 __far: limit scope of new peephole optimization rules

The two new rules I introduced were causing some bad
regressions in the gcc.c-torture/execute/ test cases.

tkchia commented:
With my latest patches to introduce proper far pointer arithmetic (commits `2b3e0f4fd3` to `856e00ac1c`), the test result summary for the `gcc` tests is as follows:

```
# of expected passes            76014
# of unexpected failures        50
# of expected failures          111
# of unresolved testcases       22
# of unsupported tests          2665
/home/user/.local/build-ia16/build2/gcc/xgcc  version 6.3.0 (GCC) 
```

The unexpected failures are [as above](https://github.com/crtc-demos/gcc-ia16/pull/2#issuecomment-333317606).  However, my `far-addr-4.c`, which had been expected to fail -- I had expressly created it to test the semantics of far pointer wrap-around -- now passes with flying colours.

I have also beefed up `far-addr-4.c` as well as `far-addr-7.c` with some additional checks.

Thank you!

tkchia committed:
IA-16: update and correct documentation on __far arithmetic

(cherry picked from commit f45e5468d925a0e69f2ec242dcb5a1f2aed9f9d5)

tkchia committed:
IA-16 __far: bug fix: only apply (subreg ...) to a register

This should fix https://github.com/tkchia/gcc-ia16/issues/4 .

ia16_expand_weird_pointer_plus_expr (...) in gcc/config/ia16/
ia16.c was trying to take a (subreg ...) of something that
might not be a register.

(cherry picked from commit 38a0c2b3ab185c473e96686d692a22d6184ee894)

tkchia committed:
IA-16 __far: bug fix: un-overload TARGET_DELEGITIMIZE_ADDRESS

This should fix https://github.com/tkchia/gcc-ia16/issues/2
and https://github.com/tkchia/gcc-ia16/issues/5 .

(take 2)

Previously I had hacked gcc/explow.c to use the hook
TARGET_DELEGITIMIZE_ADDRESS (.) for address -> pointer
conversions.  Alas, it turns out that many parts of
GCC actually use this hook to do various address ->
_address_ conversions (see e.g. avoid_constant_pool_reference
(.)), which confuses gcc/config/ia16/ia16.c and causes the
compiler to crash.

I have thus moved to overloading a different hook, namely
TARGET_LEGITIMIZE_ADDRESS (...), which now looks at its
`mode' parameter to decide whether an address -> pointer
or a pointer -> address conversion is called for.

I also cleaned up ia16_expand_weird_pointer_plus_expr (...)
a bit.

(cherry picked from commit 0a29430437f6a283aa6e5e880542292ee93640b2)

tkchia committed:
IA-16 __far: bug fix: invent new macro to convert addr. <-> ptr.

This should fix https://github.com/tkchia/gcc-ia16/issues/6 .

Rather than overload the TARGET_ADDR_SPACE_LEGITIMIZE_ADDRESS
hook to do far pointer -> address and far address -> pointer
conversions, I have invented (and documented) a new macro,
TARGET_ADDR_SPACE_CONVERT_WEIRD_MEMORY_ADDRESS (...), to do
the actual conversions.

There is now also a (nontrivial) implementation of
TARGET_ADDR_SPACE_LEGITIMIZE_ADDRESS which does what the GCC
gurus probably intended it to do: massage an address
expression into a valid machine addressing mode.

(cherry picked from commit ce8af144fe9f0bcf7eef731d544679ab4e937d8f)

tkchia committed:
IA-16 __far: bug fix: weird ptr. arith. func. now groks VOIDmode

This should fix https://github.com/tkchia/gcc-ia16/issues/7 .

(cherry picked from commit 6cfd2ca5fb93aadf16cb09e23654dd47c727872d)

tkchia committed:
IA-16 __far: added (failing) test for far array -> ptr. cast bug

See https://github.com/tkchia/gcc-ia16/issues/3 .

(cherry picked from commit f264e0344921476365f7a4c07a5d8b40af0ff204)

tkchia committed:
IA-16 __far: bug fix: keep addr. sp. info in building array type

This should fix https://github.com/tkchia/gcc-ia16/issues/3 .

Before this fix, gcc/c/c-typeck.c was creating array types
but without copying the address space information of their
element types.  This led to a compiler crash when trying to
cast a far array to a far pointer.

Note that this issue (and fix) might also affect other GCC ports
that support named address spaces.

(cherry picked from commit 1de0fcbabb3737fb43bb870e9cc4243b239f6ea5)

tkchia committed:
IA-16 __far: use proper ptr. arith. to add struct field offsets

A new test case gcc/testsuite/gcc.target/ia16/far-addr-16.c
has been added.

Also, the macro TARGET_EXPAND_WEIRD_POINTER_PLUS_EXPR now
has a somewhat different meaning: it accepts low-level `rtx'
expressions rather than high-level `tree' expressions, and
it no longer takes a `modifier' argument.

(cherry picked from commit 57fd48a1279dc7f22452a13226a01eb199ea89b1)

tkchia committed:
IA-16 __far: fix gremlins around new pointer_plus_constant (...)

(cherry picked from commit 1d39fdc8213a97a5c2ee5f4cd99b7f5fdd02b13a)

tkchia committed:
IA-16 __far: more gremlins, in ia16_as_legitimize_address (...)

(cherry picked from commit 834da48c3ba465473224f16daabfba0463cc12ac)

tkchia committed:
IA-16 __far: bug fix: tree-ssa-loop-ivopts.c special case
was handled in the wrong place

This should fix https://github.com/tkchia/gcc-ia16/issues/8 .

(cherry picked from commit cfabdd536019d69dba3d5919d0d669b3763716de)

tkchia committed:
IA-16: bug fix: earlier bug fix made GCC crash building newlib

The earlier fix to https://github.com/tkchia/gcc-ia16/issues/3
was adding `const' and `volatile' qualifiers to array types,
but composite_type () does not like qualifiers on array types
(_except_ for address space information).

(cherry picked from commit 8d5fc3f84cc37417a40c1687f0b7c53dc8b70d9f)

tkchia committed:
IA-16 __far: another gremlin

(cherry picked from commit f90953c4ae7ce8cc6728506246751603903feefd)

tkchia committed:
IA-16 __far: bug fix: constrain seg. override reg. the right way

This should fix https://github.com/tkchia/gcc-ia16/issues/9 .

(cherry picked from commit d956be236b67a7c6863b3ad8c46f8c2378109241)

tkchia committed:
IA-16: use "dg-do assemble" for far pointer test cases

This ensures that the assembly output is valid; as a bonus,
`--save-temps' lets us see what the assembly output for
each test case _is_.

(cherry picked from commit f6fc18c7532381d5fe1d2afca98c4986de39f0df)

tkchia committed:
IA-16: add test case for composite types involving far arrays
(gcc/testsuite/gcc.target/ia16/far-addr-19.c)

(cherry picked from commit 27635f7781d21368c42bae25ffbe9837d5f003f6)

tkchia committed:
IA-16: tell ivopts not to create non-original i.v. from far ptr.

This should fix https://github.com/tkchia/gcc-ia16/issues/10 .

Before this patch, the induction variable optimization
(`-fivopts') would try to fabricate an integer-typed
induction variable from a far pointer.

As 32-bit far pointers behave differently from 32-bit
integers --- especially after I implemented old-style
pointer arithmetic --- this means the new variable cannot be
properly "unified" with the original pointer variable.

Thus the "optimization" actually backfires in such a case
--- producing really bad code --- and may sometimes even
crash the compiler.

(cherry picked from commit 39987e7a91b72ac10f808204b02a47cacdba1853)

tkchia committed:
IA-16: improve gcc.info documentation on far pointers

  * Document difference between N1275 syntax and classical
    DOS compiler syntax --- as pointed out by Bart Oldeman,
    https://github.com/tkchia/gcc-ia16/issues/11 .

  * Correct note on far function pointers.

  * Add IA-16 to list of targets with named address spaces.

(cherry picked from commit 3c79fd7bc3d5a834491714fb0082a0f985e6ba83)

tkchia committed:
IA-16 __far: `-Wall' now warns about `char __far *p, *q;'

See https://github.com/tkchia/gcc-ia16/issues/11 .

There really needs to be a test case to check if the warning
is working properly.

(cherry picked from commit 88872f42ca5d21278134556ece50e028167d207b)

tkchia committed:
IA-16 __far: add 3 test cases for `char __far *p, *q;' warning

This should fix https://github.com/tkchia/gcc-ia16/issues/11
(finally).

(cherry picked from commit 3b13fc9ccbe11c2ea0f33cda33b23f319b6f2218)

tkchia committed:
IA-16 __far: further fix gcc/lra-constraints.c's process_addr_1 ()

This should fix https://github.com/tkchia/gcc-ia16/issues/14 .
I added two test programs for this bug.

A test case for https://github.com/tkchia/gcc-ia16/issues/12
--- which is _not_ yet fixed --- has also been added.

(cherry picked from commit f25dcbc5c13ae84054e04c8d28589a9a5b6a0e4d)

tkchia committed:
IA-16: bug fix: peephole on `movw %es, %bx; movw (%bx), %bx'

One of the peephole rules was wrongly optimizing `movw %es,
%bx; movw mem, %bx' into `movw mem, %bx' even when mem was
`(%bx)'.  The correct thing to do was to check %bx for
liveness before the `movw mem, ...'

This fixes a regression in gcc/testsuite/gcc.c-torture/
execute/bitfld-3.c.  The regression showed up
intermittently, so I initially misdiagnosed it as being
caused by changes to the register file.

(cherry picked from commit 4a587b427b54aa8b7150fd578be8960bb9578632)

tkchia commented:
Hello @reenigne,

I think I have found a workable way to make `%ds` available for register allocation while preserving the current ABI (assume `%ds` == `%ss` on function entry and restore this on exit).  I have implemented an initial version of this in [the](https://github.com/tkchia/gcc-ia16/commit/b6abfefdbaea82b620e703f745d8353368f43d5f) [latest](https://github.com/tkchia/gcc-ia16/commit/b58280cf1bb39edfbbc6b4540584f0ea48127774) [commits](https://github.com/tkchia/gcc-ia16/commit/8a25d13ff906dfc12158dfad0d028e18a33396ef) on my own development branch, and I hope to possibly file a separate pull request for it.

In outline, what my code does is
  * make `%ds` available in the register file as a _callee-clobbered_ register (`%ds` is considered to be "clobbered" to a particular value upon each subroutine call);
  * add a `pushw %ss; popw %ds` sequence before each subroutine call, and within each function epilogue; then
  * using a `TARGET_MACHINE_DEPENDENT_REORG` hook, analyze each function's control and data flow, and remove any unneeded `pushw %ss; popw %ds` sequences.

Currently, any function which modifies `%ds` will always access variables in the generic space via `%ss` (e.g. `mov %ss:(%bx), %al`).  It should be possible to optimize this too, though I have not yet done so.

Thank you!
